src/ring_theory/dedekind_domain.lean

```diff
@@ -891,3 +892,58 @@ instance integral_closure.is_dedekind_domain_fraction_ring
 integral_closure.is_dedekind_domain A (fraction_ring A) L
 
 end is_integral_closure
+
+section is_dedekind_domain
+
+variables {T : Type*} [integral_domain T] [is_dedekind_domain T] (I J : ideal T)
+open_locale classical
+open multiset unique_factorization_monoid ideal
+
+lemma prod_factors_eq_ideal {I : ideal T} (hI : I ≠ ⊥) : (factors I).prod = I :=
+associated_iff_eq.1 (factors_prod hI)
+
+lemma factors_prod_factors_eq_factors {α : multiset (ideal T)} (h₁ : ⊥ ∉ α)
+  (h₂ : ∀ p ∈ α, prime p) : factors α.prod = α :=
+by {simp_rw [← multiset.rel_eq, ← associated_eq_eq], exact prime_factors_unique
+(prime_of_factor) h₂ (factors_prod (prod_ne_zero h₁))}
+
+lemma count_le_of_ideal_ge {I J : ideal T} (h : I ≤ J) (hI : I ≠ ⊥) (K : ideal T) :
+  count K (factors J) ≤ count K (factors I) :=
+by apply le_iff_count.1 ((dvd_iff_factors_le_factors (ne_bot_of_ne_bot_le h hI) hI).1
+  (dvd_iff_le.2 h))
+
+lemma zero_not_mem_inf_factors (I J : ideal T) : ⊥ ∉ factors I ⊓ factors J :=
+begin
+  by_contra hcontra,
+  rw [multiset.inf_eq_inter, mem_inter] at hcontra,
+  exact false_of_ne (prime.ne_zero (prime_of_factor (0 : ideal T) hcontra.left)),
+end
+
+lemma sup_eq_prod_inf_factors (hI : I ≠ ⊥) (hJ : J ≠ ⊥) : I ⊔ J = (factors I ⊓ factors J).prod :=
+begin
+  have H : factors (factors I ⊓ factors J).prod = factors I ⊓ factors J,
+  { apply factors_prod_factors_eq_factors (zero_not_mem_inf_factors I J),
+    intros p hp,
+    rw [multiset.inf_eq_inter, mem_inter] at hp,
+    exact prime_of_factor p hp.left },
+  apply le_antisymm,
+  { rw [sup_le_iff, ← dvd_iff_le, ← dvd_iff_le],
+    split,
+    {rw [dvd_iff_factors_le_factors (prod_ne_zero (zero_not_mem_inf_factors I J)) hI, H],
+      exact inf_le_left },
+    { rw [dvd_iff_factors_le_factors (prod_ne_zero (zero_not_mem_inf_factors I J)) hJ, H],
+      exact inf_le_right } },
+  { rw [← dvd_iff_le, dvd_iff_factors_le_factors, factors_prod_factors_eq_factors
+    (zero_not_mem_inf_factors I J), le_iff_count],
+    { intro a,
+    rw [multiset.inf_eq_inter, multiset.count_inter],
+    exact le_min (count_le_of_ideal_ge le_sup_left hI a)
+      (count_le_of_ideal_ge le_sup_right hJ a) },
+    { intros p hp,
+      rw [multiset.inf_eq_inter, mem_inter] at hp,
+      exact prime_of_factor p hp.left },
+    { exact ne_bot_of_ne_bot_le le_sup_left hI },
+    { exact prod_ne_zero (zero_not_mem_inf_factors I J) } },
```

Comment:
Do you have another use in mind for `zero_not_mem_inf_factors`? It seems like it would only be relevant in the proof that follows, as an argument to `prod_ne_zero`. So I inlined that combination using the `have` tactic, and generalized `zero_not_mem_inf_factors` to a statement that holds on any UFM.

I also replaced the `⊓` operation on multisets with the intersection, which saves a couple of rewrites. Typically if a `simp` lemma like `multiset.inf_eq_inter` applies to the theorem statement, then you should indeed `simp` it.
```suggestion
-- can be moved to ring_theory/unique_factorization_monoid.lean
lemma zero_not_mem_factors {M : Type*} [comm_cancel_monoid_with_zero M] [nontrivial M]
  [normalization_monoid M] [unique_factorization_monoid M] (x : M) : (0 : M) ∉ factors x :=
λ h, prime.ne_zero (prime_of_factor _ h) rfl

lemma sup_eq_prod_inf_factors (hI : I ≠ ⊥) (hJ : J ≠ ⊥) : I ⊔ J = (factors I ∩ factors J).prod :=
begin
  have H : factors (factors I ∩ factors J).prod = factors I ∩ factors J,
  { apply factors_prod_factors_eq_factors,
    intros p hp,
    rw mem_inter at hp,
    exact prime_of_factor p hp.left },
  have := ((factors I ∩ factors J).prod_ne_zero_of_prime
    (λ _ h, prime_of_factor _ (multiset.mem_inter.1 h).1)),
  apply le_antisymm,
  { rw [sup_le_iff, ← dvd_iff_le, ← dvd_iff_le],
    split,
    { rw [dvd_iff_factors_le_factors this hI, H],
      exact inf_le_left },
    { rw [dvd_iff_factors_le_factors this hJ, H],
      exact inf_le_right } },
  { rw [← dvd_iff_le, dvd_iff_factors_le_factors, factors_prod_factors_eq_factors, le_iff_count],
    { intro a,
      rw multiset.count_inter,
      exact le_min (count_le_of_ideal_ge le_sup_left hI a)
        (count_le_of_ideal_ge le_sup_right hJ a) },
    { intros p hp,
      rw mem_inter at hp,
      exact prime_of_factor p hp.left },
    { exact ne_bot_of_ne_bot_le le_sup_left hI },
    { exact this } },
```

Reply:
I agree, it's best to remove `zero_not_mem_inf_factors` (I wasn't very sure in the first place if it was worth stating as a lemma or not, and I can't think of any others uses of it for the time being !)